libtransmission/peer-mgr.cc

```diff
@@ -222,38 +263,26 @@ class tr_swarm
         auto const lock = unique_lock();
 
         is_running = false;
-        removeAllPeers();
+        remove_all_peers();
         outgoing_handshakes.clear();
     }
 
-    void removePeer(tr_peer* peer)
+    void remove_peer(tr_peerMsgs* peer)
     {
-        auto const lock = unique_lock();
-
-        auto* const peer_info = peer->peer_info;
-        TR_ASSERT(peer_info != nullptr);
-
         if (auto iter = std::find(std::begin(peers), std::end(peers), peer); iter != std::end(peers))
         {
             peers.erase(iter);
+            close_peer(peer);
         }
-
-        --stats.peer_count;
-        --stats.peer_from_count[peer_info->from_first()];
-
-        TR_ASSERT(stats.peer_count == peerCount());
-
-        delete peer;
     }
 
-    void removeAllPeers()
+    void remove_all_peers()
     {
-        auto tmp = peers;
-
-        for (auto* peer : tmp)
+        for (auto* peer : peers)
```

Comment:
...hm. I know I suggested this, but the change I suggested has introduced temporarily-dangling pointers in `peers` and now I'm wondering if there are any code paths where another thread might walk that vector.

If there's nothing in `closePeer()` that requires the pointer to be present in `peers`, probably safer to do something like

```c++
void remove_all_peers()
{
    auto tmp = Peers{};
    std::swap(peers, tmp);
    for (auto* peer : tmp)
    {
        removePeer(peer);
    }
}
````

This introduces a new vector temporary; but since we're just swapping, the cost should be essentially free.

Alternately, if you don't like this then consider reverting to the reverse iterator implementation :man_shrugging:

Reply:
> now I'm wondering if there are any code paths where another thread might walk that vector.

I found that the GTK code calls `tr_torrentStat()` in a few places, which calls `tr_peerMgrGetDesiredAvailable()` that iterates through `peers`. Do you have an idea if these code are run in another thread?

- gtk/Session.cc:598
- gtk/Torrent.cc:205
- gtk/Dialogs.cc:42

If they are, since `remove_all_peers()` was not thread-safe in the first place, is this a gtk bug?

Reply:
Found another similar usage with the Mac client calling `tr_torrentPeers()`, which then calls `tr_peerMgrPeerStats()` that loops through `peers`.

Reply:
Yes, this is risky behavior.

I tried some crude mutex locking in e4b480ecd45f79ff302a60ad2347a10e8dd7d4db to make it safer, but the performance tradeoff was too high so it was reverted in adf5051d467f0384e046d5c6e37092a04c9baa58. See #4571.

A full fix for this would probably be scope creep for this PR and the risk is definitely not new to this PR, but we still probably ought to make `remove_all_peers()` safer either by going back to the reverse_iterator approach or the swap() alternative.

Reply:
Right, roger that.

Reply:
Opted for the tmp vector because it is the most thread-safe implementation right now.

8848fd6